### PR TITLE
CRUD backend: Add volumesnapshot and get_pvc functions

### PR DESCRIPTION
**Pull request by DavidSpek**
Part of: https://github.com/kubeflow/kubeflow/issues/5523

This PR adds functions for creating volume snapshots to the common backend code for the CRUD web apps. It also adds a function to get an existing PVC.

/cc @kimwnasptd @thesuperzapper @StefanoFioravanzo

### CODE REVIEW

**Comment by thesuperzapper** [comment]
@DavidSpek can you rename the `snapshot.py` (and associated functions) to `volumesnapshot.py`?

File: components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/api/__init__.py [hunk 1]
```diff
@@ -7,5 +7,6 @@
 from .poddefault import *  # noqa F401, F403
 from .pvc import *  # noqa F401, F403
 from .secret import *  # noqa F401, F403
+from .snapshot import *  # noqa F401, F403
```
Comment (thesuperzapper) [inline components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/api/__init__.py:10]: This is wrong

**Comment by kimwnasptd** [comment]
@DavidSpek I understand this is part of a bigger effort of integrating VolumeSnapshots to the existing web apps and UX. 

Could you also create an issue, expose the top-down approach and link to this PR as a first step for that effort? We can discuss and shape the overarching goal and what makes sense to implement there and keep it as a tracking issue.

**Comment by DavidSpek** [comment]
@kimwnasptd I'll create a tracking issue with my ideas in a moment, but I don't think that should block this PR.

**Comment by thesuperzapper** [comment]
@kimwnasptd this PR is very general, (and useful outside any specific use-case), so unless you see any specific issues, let's just get this merged, as its good functionality for the common backend.

**Comment by kimwnasptd** [comment]
> @kimwnasptd this PR is very general

I know, but I want us to start exposing our plans more in issues first so that it's more visible to contributors what are our long term goals. This will allow both us and other people to be able to track the progress of a feature, even if it's broken into many smaller PR [which of course is good!].

That's why I'd like us to have a tracking issue to discuss the long term design, which this PR is the first step towards.

**Comment by DavidSpek** [comment]
@kimwnasptd In the working group meetings it was discussed and agreed upon many times that creating a tracking issue with a proposal for implementing small code changes/features is not necessary or productive. All this is doing is blocking the development of fixes and features users have asked for. Writing a detailed tracking issue or proposal feels like a waste of time as it will likely not get looked at if history is anything to go by. I've probably spent 10x the time, if not more, trying to get PRs looked at and merged than the actual coding took. As such, it isn't surprising that most people that create a PR for this repository let it go stale since the approvers are unresponsive. Needless to say, this is not a sustainable way to run a project and it is a clear indication that actions need to be taken with regards to the approvers of the components in this repository (as stated in the [Contributing to Kubeflow](https://www.kubeflow.org/docs/about/contributing/#quirks-of-the-process) section on the website), as well as the lack of diversity amongst the leads of this working group.

**Comment by DavidSpek** [comment]
@kubeflow/wg-notebooks-leads Can this PR with boilerplate code be merged so that we can move forward?

**Comment by kimwnasptd** [comment]
The web apps don't interact with CronJobs and thus we don't have API helper functions for CronJobs. The web apps don't interact with VolumeSnapshots thus there's currently no need to add support for volume snapshots. This is also the case for Jobs, ReplicaSets, Endpoints, Ingress, LimitRangers etc.

If you'd like to expand the apps to support different use cases and objects then we will need a more detailed proposal, which can be tracked in an issue as a first step.

**Comment by DavidSpek** [comment]
@kimwnasptd I have created detailed proposals in the past, which didn't weren't looked at and I ended up just wasting a lot of my time on. It's also very difficult to decide what a proposal should look like, since you don't respond when I reach out to you.

As I also explained above, during the meetings there was consensus that a proposal would not be necessary for small feature improvements. I would consider allowing a user to create a new PVC from an existing snapshot a small feature. And since this needs to be added to 2 web apps (jupyter and volumes) it makes sense to first have the necessary backend code in place.

I'm seriously beginning to wonder if this is being blocked to protect the interests of Arrikto as volume snapshots are the main selling point of Rok, as 4 out of 5 of the working group leads are from Arrikto. This highlights the importance of neutral governance within working groups, as the interests of a single company should not be able to block Kubeflow development.

**Comment by kimwnasptd** [comment]
> @kimwnasptd I have created detailed proposals in the past, which didn't weren't looked at and I ended up just wasting a lot of my time on.

Please point me to the proposals I’ve missed, there’s always the chance I’ve missed something. However we try to reply to everything. Also, I can understand that writing a feature proposal and documentation can be time consuming but this is what makes the project long-term sustainable.

Having written down proposals can help us all get in sync with what we want to implement, discuss technical details and agree on the implementation.

> It's also very difficult to decide what a proposal should look like, since you don't respond when I reach out to you.

You don't need to have a 100% thought out proposal from the get go. The current feature-issue-template can be a very good place to describe the features you'd like to see/implement and propose a design. Then we can use that issue to evaluate how we could implement such a feature.

> I would consider allowing a user to create a new PVC from an existing snapshot a small feature. And since this needs to be added to 2 web apps...

A change that extends both the backend and the frontend of 2 web apps and changes user facing UIs is by no means a small feature. There are a lot of things to discuss, from avoiding code duplication to deciding on how the UI should look like as well as API changes to accommodate the common code.

**Comment by DavidSpek** [comment]
How is this for a proposal for this feature: https://github.com/kubeflow/kubeflow/issues/5523

**Pull request closed without merging**